### DISCUSSION TITLE
Enable Link-Time Optimization (LTO)

### DISCUSSION THREAD

**Discussion by zamazan4ik** [jj-vcs/jj#4463] [Ideas]
Hi!

I noticed that in the `Cargo.toml` file Link-Time Optimization (LTO) for the project is not enabled. I suggest switching it on since it will reduce the binary size (always a good thing to have) and will likely improve the application's performance a bit (in this case, it's not critical at all but anyway).

I suggest enabling LTO only for the Release builds so as not to sacrifice the developers' experience while working on the project since LTO consumes an additional amount of time to finish the compilation routine. If you think that a regular Release build should not be affected by such a change as well, then I suggest adding an additional `dist` or `release-lto` profile where additionally to regular `release` optimizations LTO also will be added. Such a change simplifies life for maintainers and others interested in the project persons who want to build the most performant version of the application. Using ThinLTO also should help.

I already [found](https://github.com/martinvonz/jj/commit/7acdc8f01111f800cc49648d4ee00a8d06116747) a commit that has the following words:

> It *might* be worth exploring if something like `lto=thin` could improve performance since it didn't meaningfully impact compile time or size any further, but that's TBD.

Maybe it's time to reconsider the decision ;)

Thank you.

**Comment by martinvonz** [comment]
I think @thoughtpolice knows most about this. Feel free to experiment with it and share the results. IMO, if it's a clear win in binary size and not a loss in runtime performance and at most a small loss in build speed, it seems reasonable to do.

**Reply by zamazan4ik** [reply]
I performed some quick benchmarks on the latest commit (`main` branch, `05c6d62c681598697d345e18f802f5e601dac1f7` commit, Fedora 40, Rustc 1.81):

* Release binary size: 21 Mib, build time: 1m 33s
* Release + `lto = true`: 19 Mib, build time: 2m 40s
* Release + `lto = "thin"`: 22 Mib, build time: 1m 39s

What is the recommended way to measure the performance of the tool? I also want to compare performance between different LTO versions
